src/BasicInputs/DatetimeInput.tsx

```diff
@@ -1,27 +1,130 @@
-import {
-  type ComponentPropsWithoutRef,
-  type ElementRef,
-  forwardRef,
-} from 'react';
+'use client';
+
+import { forwardRef } from 'react';
+
+import { add, format } from 'date-fns';
+import { Calendar as CalendarIcon } from 'lucide-react';
 
-import type { FormControlChildProps } from '@/Form';
-import { Input } from '@/lib/components/ui/input';
 import { cn } from '@/lib/utils';
 
-export const DatetimeInput = forwardRef<
-  ElementRef<typeof Input>,
-  FormControlChildProps & ComponentPropsWithoutRef<typeof Input>
->(({ value = '', className, ...inputProps }, ref) => {
-  if (typeof value !== 'string') {
-    throw new Error('value must be of type string.');
-  }
-  return (
-    <Input
-      type="datetime-local"
-      value={value?.substring(0, 19) ?? ''}
-      className={cn('dark:[color-scheme:dark]', className)}
-      ref={ref}
-      {...inputProps}
-    />
-  );
-});
+import { Button } from '../lib/components/ui/button';
+import { Calendar } from '../lib/components/ui/calendar';
+import {
+  Popover,
+  PopoverContent,
+  PopoverTrigger,
+} from '../lib/components/ui/popover';
+import {
+  type Granularity,
+  TimePicker12h,
+  TimePicker24h,
+} from '../lib/components/ui/time-picker';
+
+type DatetimeInputRef = {
+  value?: string;
+} & Omit<HTMLButtonElement, 'value'>;
+
+export interface DatetimeInputProps {
+  value?: string | null;
+  onChange?: (date: string | undefined | null) => void;
+  granularity?: Granularity;
+  hourCycle?: 12 | 24;
+}
+
+export const DatetimeInput = forwardRef<DatetimeInputRef, DatetimeInputProps>(
+  (
+    { value, onChange, granularity = 'Second', hourCycle = 12, ...props },
+    ref,
+  ) => {
+    const date: Date | undefined = value ? new Date(value) : undefined;
+
+    /**
+     * carry over the current time when a user clicks a new day
+     * instead of resetting to 00:00
+     */
+    const handleSelect = (newDay: Date | undefined) => {
+      if (!newDay || !date) return onChange?.(newDay?.toISOString());
+
+      const diff = newDay.getTime() - date.getTime();
+      const diffInDays = diff / (1000 * 60 * 60 * 24);
+      const newDateFull = add(date, { days: Math.ceil(diffInDays) });
+      onChange?.(newDateFull.toISOString());
+    };
```

Comment:
You can reuse onDateChange within here. This makes it so that any error/validation handling is just from that fn. 

Basically every instance of onChange within this can now just be passed the date object since onDateChange converts it to a datetime string | null for us.